### PR TITLE
Return ERefs rather than Promises when possible

### PR DESCRIPTION
**Pull request by erights**
Low priority. Not needed for MN1. But nice cleanups.

Functions that incidentally return a `Promise<T>` should also be able to return a `T`, and thus should be typed `ERef<T>`. Any function that currently type checks against the `Promise<T>` return type should continue to type check. But callers that were counting on it being a promise are encouraged by the type system to be more tolerant. This reserved refactoring possibilities for the implementation. Doing this revealed places where we were directly using `.then` when we should have been using `E.when`.

Unfortunately, TS does not allow an `async` function to be declared as returning an `ERef`. It insists on a `Promise`. Thus, we are forced to type these functions in a way that allows callers to rely on it being a promise, precluding attractive refactorings.

### CODE REVIEW

File: packages/ERTP/test/unitTests/test-issuerObj.js [hunk 1]
```diff
@@ -104,7 +104,7 @@ test('issuer.makeEmptyPurse', async t => {
   const performWithdrawal = () => purse.withdraw(fungible837);
 
   const checkWithdrawal = async newPayment => {
-    issuer.getAmountOf(newPayment).then(amount => {
+    E.when(issuer.getAmountOf(newPayment), amount => {
```
Comment (mhofman) [inline packages/ERTP/test/unitTests/test-issuerObj.js:107]: There was apparently a double error here? Shouldn't the result of `E.when()` be `await`ed?

File: packages/ERTP/test/unitTests/test-issuerObj.js [hunk 2]
```diff
@@ -255,6 +255,8 @@ test('issuer.burn', async t => {
     AmountMath.isEqual(burntBalance, AmountMath.make(brand, 837n)),
     `entire minted payment was burnt`,
   );
+  // @ts-expect-error ERef arg should be compat to Promise or PromiseLike
```
Comment (mhofman) [inline packages/ERTP/test/unitTests/test-issuerObj.js:258]: So this actually brings up an issue with `ERef` and the expectation that the caller should be able to handle a synchronous result. Most code will not be able to handle a synchronous exception, and I'd argue it shouldn't.

After going back and forth on what I'd expect to happen, the ava implementation in this case does what I think is right, which is to fail the test if the function throws synchronously. It also fails if the function returns a value which is not a promise. https://github.com/avajs/ava/blob/v3.15.0/lib/assert.js#L588-L606

So I'd argue that technically `ERef` is not compatible with `PromiseLike` in general, but that in this case it is, because only a rejecting `PromiseLike` result will pass the test.

File: packages/ERTP/test/unitTests/test-issuerObj.js [hunk 3]
```diff
@@ -411,7 +421,7 @@ test('issuer.combine array of promises', async t => {
   harden(paymentsP);
 
   const checkCombinedResult = paymentP => {
-    issuer.getAmountOf(paymentP).then(pAmount => {
+    E.when(issuer.getAmountOf(paymentP), pAmount => {
```
Comment (mhofman) [inline packages/ERTP/test/unitTests/test-issuerObj.js:424]: Should this be `await`ed as well?

File: packages/notifier/src/storesub.js [hunk 4]
```diff
@@ -60,10 +60,17 @@ export const makeStoredSubscription = (
 
   if (storageNode) {
     // Start publishing the source.
-    observeIteration(subscription, {
-      updateState: publishValue,
-      finish: publishValue,
-    }).catch(fail);
+    E.when(
```
Comment (mhofman) [inline packages/notifier/src/storesub.js:63]: Handle promise ?

**Review by mhofman — COMMENTED** [review]
The replacements from `.then()` to `E.when()` look fine, but do seem to highlight some previously missing unhandled promises. I'm not sure why they weren't warned about before.

I am a little worried about changing `Promise<void>` to `ERef<void>` since the `void` type has a special meaning of no return type, and is thus contagious. In practice an `ERef<void>` is equivalent to `void` (with the Promise part of the union being dropped), which would theoretically make `await`ing the result appear unnecessary to the type system. See https://github.com/microsoft/TypeScript/issues/42709.

Similarly `ERef<any>` will result in `any` since it's contagious too, but at least TS will not complain when `await`ing `any`.

Finally, my biggest concern is with error handling for `ERef` returning functions. These functions would imply that they may throw synchronously, which would make code like `E.when(foo(), handleFooResult)` problematic.

To summarize, I'm fairly uncomfortable with switching to `ERef` for return types in such wholesale fashion.

**Comment by erights** [comment]
Closing. We may eventually scavenge some useful code from the carcass of this PR.
